src/encodings.c

```diff
@@ -428,85 +434,61 @@ void encodings_init(void)
 	{
-		menu_westeuro = gtk_menu_new();
-		item_westeuro = gtk_menu_item_new_with_mnemonic(_("_West European"));
-		gtk_menu_item_set_submenu(GTK_MENU_ITEM(item_westeuro), menu_westeuro);
-		gtk_container_add(GTK_CONTAINER(menu[k]), item_westeuro);
-		gtk_widget_show_all(item_westeuro);
-
-		menu_easteuro = gtk_menu_new();
-		item_easteuro = gtk_menu_item_new_with_mnemonic(_("_East European"));
-		gtk_menu_item_set_submenu(GTK_MENU_ITEM(item_easteuro), menu_easteuro);
-		gtk_container_add(GTK_CONTAINER(menu[k]), item_easteuro);
-		gtk_widget_show_all(item_easteuro);
-
-		menu_eastasian = gtk_menu_new();
-		item_eastasian = gtk_menu_item_new_with_mnemonic(_("East _Asian"));
-		gtk_menu_item_set_submenu(GTK_MENU_ITEM(item_eastasian), menu_eastasian);
-		gtk_container_add(GTK_CONTAINER(menu[k]), item_eastasian);
-		gtk_widget_show_all(item_eastasian);
-
-		menu_asian = gtk_menu_new();
-		item_asian = gtk_menu_item_new_with_mnemonic(_("_SE & SW Asian"));
-		gtk_menu_item_set_submenu(GTK_MENU_ITEM(item_asian), menu_asian);
-		gtk_container_add(GTK_CONTAINER(menu[k]), item_asian);
-		gtk_widget_show_all(item_asian);
-
-		menu_middleeast = gtk_menu_new();
-		item_middleeast = gtk_menu_item_new_with_mnemonic(_("_Middle Eastern"));
-		gtk_menu_item_set_submenu(GTK_MENU_ITEM(item_middleeast), menu_middleeast);
-		gtk_container_add(GTK_CONTAINER(menu[k]), item_middleeast);
-		gtk_widget_show_all(item_middleeast);
-
-		menu_utf8 = gtk_menu_new();
-		item_utf8 = gtk_menu_item_new_with_mnemonic(_("_Unicode"));
-		gtk_menu_item_set_submenu(GTK_MENU_ITEM(item_utf8), menu_utf8);
-		gtk_container_add(GTK_CONTAINER(menu[k]), item_utf8);
-		gtk_widget_show_all(item_utf8);
-
-		/** TODO can it be optimized? ATM 3782 runs at line "if (encodings[j].group ...)" */
+		GSList *group = NULL;
+		GtkWidget *submenus[GEANY_ENCODING_GROUPS_MAX];
+		guint i;
+		gint orders[GEANY_ENCODING_GROUPS_MAX] = { 0 };
+		guint n_added = 0;
+
 		for (i = 0; i < GEANY_ENCODING_GROUPS_MAX; i++)
 		{
-			order = 0;
-			switch (i)
+			if (! groups[i]) /* NONE */
+				submenus[i] = menu[k];
+			else
 			{
-				case WESTEUROPEAN: submenu = menu_westeuro; group_size = 9; break;
-				case EASTEUROPEAN: submenu = menu_easteuro; group_size = 14; break;
-				case EASTASIAN: submenu = menu_eastasian; group_size = 14; break;
-				case ASIAN: submenu = menu_asian; group_size = 9; break;
-				case MIDDLEEASTERN: submenu = menu_middleeast; group_size = 7; break;
-				case UNICODE: submenu = menu_utf8; group_size = 8; break;
-				default: submenu = menu[k]; group_size = 1;
+				GtkWidget *item = gtk_menu_item_new_with_mnemonic(_(groups[i]));
+				submenus[i] = gtk_menu_new();
+				gtk_menu_item_set_submenu(GTK_MENU_ITEM(item), submenus[i]);
+				gtk_container_add(GTK_CONTAINER(menu[k]), item);
+				gtk_widget_show_all(item);
 			}
+		}
 
-			while (order < group_size)	/* the biggest group has 13 elements */
+		/** TODO can it be optimized? ATM 882 runs at line "if (encoding->group ...)" */
```

Comment:
well, it's still not great so maybe.  And yeah, fixed the reference that I changed at some point but changed it back later.